substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/InstrumentFeature.java

```diff
@@ -0,0 +1,118 @@
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.svm.hosted;
+
+import com.oracle.svm.core.PreMainSupport;
+import com.oracle.svm.core.feature.AutomaticallyRegisteredFeature;
+import com.oracle.svm.core.feature.InternalFeature;
+import com.oracle.svm.core.option.HostedOptionKey;
+import com.oracle.svm.core.option.LocatableMultiOptionValue;
+import com.oracle.svm.core.util.VMError;
+import com.oracle.svm.hosted.reflect.ReflectionFeature;
+import jdk.graal.compiler.options.Option;
+import org.graalvm.nativeimage.ImageSingletons;
+import org.graalvm.nativeimage.hosted.Feature;
+
+import java.lang.instrument.Instrumentation;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This feature supports instrumentation in native image.
+ */
+@AutomaticallyRegisteredFeature
+public class InstrumentFeature implements InternalFeature {
+    private ClassLoader cl;
+    private PreMainSupport preMainSupport;
+
+    public static class Options {
+        @Option(help = "Specify premain-class list. Multiple classes are separated by comma, and order matters.")//
+        public static final HostedOptionKey<LocatableMultiOptionValue.Strings> PremainClasses = new HostedOptionKey<>(null);
+
+    }
+
+    /**
+     * {@link ReflectionFeature} must come before this feature, because many instrumentation methods
+     * are called by reflection, e.g. premain.
+     */
+    @Override
+    public List<Class<? extends Feature>> getRequiredFeatures() {
+        return List.of(ReflectionFeature.class);
+    }
+
+    @Override
+    public void afterRegistration(AfterRegistrationAccess access) {
+        FeatureImpl.AfterRegistrationAccessImpl a = (FeatureImpl.AfterRegistrationAccessImpl) access;
+        cl = a.getImageClassLoader().getClassLoader();
```

Comment:
I'm not sure this is still required, but I had to provide an agent JAR twice as `-J-javaagent` and `-cp` for NI build configurations in agent JARs to be picked up by the NI builder. So just checking, users can provide an agent JAR only once as `-J-javaagent` for their premain classes to be registered?

Reply:
In this PR, we are not automatically picking agent from `-cp` or `-J-javaagent`.  It will be handled in the following PRs. But I'm sure user doesn't need to provide agent jar multiple times.